### PR TITLE
refactor: update packaging content + better site infra

### LINKED ISSUE #196: Mention of conference about packaging
That may be a new source of resources and presentations to learn from. First one will be on Nov 2021
https://packaging-con.org/

### LINKED ISSUE #237: Check packaging information is up to date on notes and exercises
Packaging explanation has been updated to use `pyproject.toml` though some exercises may not have it([from week 7](https://github.com/UCL-COMP0233-22-23/RSE-Classwork/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Aweek07).
We need to update the explanations, add links to documentation such as [python's packaging guide](https://packaging.python.org/en/latest/) and [pyopensci's packaging guide](https://www.pyopensci.org/python-package-guide/). Maybe even to teach [`hatch`](https://www.pyopensci.org/blog/pyos-community-news-aug-2024.html).

### PR DESCRIPTION
**Pull request by Saransh-cpp**
Closes #237
Closes #196

Packaging contents update:
- More resources (packaging guides, cookie templates, more backend descriptions)
- Use `hatch` instead of `setuptools`
- Follow a different (`src`) directory structure (match it with scientific python's guidelines)
- Some more information about VCS versioning, optional dependencies, ...

Site infra updates:
- Get rid of old and archived actions
- Use the new GitHub sites build instead of `gh-pages`
- Upload the generated site as an artifact to preview it locally for validation

### CODE REVIEW

File: .github/workflows/build_site.yml [hunk 1]
```diff
@@ -31,23 +31,28 @@ jobs:
...
-        uses: helaili/jekyll-action@v2
+      - uses: ruby/setup-ruby@v1
         with:
-          target_branch: 'gh-pages'
-          build_only: true
-      - name: Build website and deploy
-        if: github.ref == 'refs/heads/main'
-        uses: helaili/jekyll-action@v2
+          ruby-version: '3.3'
+      - name: Install apt requirements
```
Comment (Saransh-cpp) [inline .github/workflows/build_site.yml:44]:
```suggestion
      - name: Install bundler
```

File: .github/workflows/build_site.yml [hunk 2]
```diff
@@ -31,23 +31,28 @@ jobs:
...
+        run: gem install bundler
+      - name: Install apt requirements
```
Comment (Saransh-cpp) [inline .github/workflows/build_site.yml:46]:
```suggestion
      - name: Build website
```

File: README.md [hunk 3]
```diff
@@ -79,6 +79,13 @@ $ docker run --rm --volume="$PWD:/srv/jekyll" --volume="$PWD/vendor/bundle:/usr/
 and open http://localhost:4000/rsd-engineeringcourse (or the link provided).
 Note that this is mounting the `bundle` directory where `act` will create them.
 
+or if you want to simply build and serve the site locally (with `ruby` and `bundle` installed):
+
+```bash
+$ bundle install
+$ bundle exec jekyll serve
+```
```
Comment (Saransh-cpp) [inline README.md:87]:
```suggestion
```

File: ch04packaging/03Packaging.ipynb.py [hunk 4]
```diff
@@ -21,43 +21,57 @@
...
-# - Language-specific tools: 
-#  - python: PyPI,
-#  - ruby: Ruby Gems, 
-#  - perl: CPAN,
-#  - R: CRAN
+#  - Language-specific tools: 
+#    - python: PyPI,
+#    - ruby: Ruby Gems, 
+#    - perl: CPAN,
+#    - R: CRAN
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:47]: Maybe we should add here some explanation and links about this. In fact, I don't know whether our intention here was to link tools (`pip`, `gem`, ...) or repositories (which all are!). So maybe we could add a section about repositories (and mention that list together with the one below without much detail), and then there mentioning the tools to query them.

File: ch04packaging/03Packaging.ipynb.py [hunk 5]
```diff
@@ -21,43 +21,57 @@
...
+# |   `-- package_name
 # |       |-- __init__.py
-# |       `-- test_python_file.py
+# |       |-- python_file.py
+# |       |-- another_python_file.py
+# `-- tests
+#     |-- fixtures
+#     |   `-- fixture_file.yaml
+#     |-- __init__.py
+#     `-- test_python_file.py
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:74]: I've never been a fun of this structure, but I know that's what all are suggesting. I don't like to have the tests far from the source as it makes it harder to follow TDD. But, if that's what all guides suggest and how the community is moving, I imagine I'll have to adapt :sweat_smile:

File: ch04packaging/03Packaging.ipynb.py [hunk 6]
```diff
@@ -94,30 +109,45 @@
...
+[tool.hatch.build.targets.sdist]
+include = [
+  "src*",
+]
 
 # %% [markdown]
-# We can now install this "package" with pip:
+# Some of the build-backends allow users to automate the package's version using VCS.
+# For instance, you might want to look into [`hatch-vcs`](https://github.com/ofek/hatch-vcs) to enable VCS versioning with `hatch`.
+# We can now install this "package" with pip (make sure `hatch` is installed):
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:150]:
```suggestion
# For instance, you might want to look into [`hatch-vcs`](https://github.com/ofek/hatch-vcs) to enable VCS versioning with `hatch`.
#
# We can now install this "package" with pip (make sure `hatch` is installed):
```

File: ch04packaging/03Packaging.ipynb.py [hunk 7]
```diff
@@ -362,21 +404,18 @@ def process():
...
 greet = "greetings.command:process"
 
-[build-system]
-requires = ["setuptools", "setuptools_scm[toml]>=6.2", "wheel"]
-
-[tool.setuptools.packages.find]
-include = ["greetings*"]
+[tool.hatch.build.targets.sdist]
+include = [
+  "greetings/",
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:409]:
```suggestion
  "src/",
```
that's at least what we have above. `greetings` or would it be `src/greetings/`?

File: ch04packaging/03Packaging.ipynb.py [hunk 8]
```diff
@@ -362,21 +404,18 @@ def process():
...
-# Add setuptools_scm if you need to generate version numbers from the git hash
-#[tool.setuptools_scm]
 
 # %% [markdown]
 # When installing the package now, pip will also install the dependencies automatically.
 
 # %% language="bash"
 # cd greetings_repo
-# pip install -e .
+# pip install .
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:418]: You don't like the `-e` or is not needed with hatch? If we are changing the library, `e` gives us (or used to) an editable library.

File: ch04packaging/03Packaging.ipynb.py [hunk 9]
```diff
@@ -468,16 +507,16 @@ def process():
 Portions of the material are taken from [Software Carpentry](http://software-carpentry.org/)
 
 # %% [markdown]
-# You may well want to formalise this using the [codemeta.json](https://codemeta.github.io/) standard or the [citation file format](http://citation-file-format.github.io/) - these don't have wide adoption yet, but we recommend it.
+# You may well want to formalise this using the [codemeta.json](https://codemeta.github.io/) standard or the [citation file format](http://citation-file-format.github.io/).
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:510]: Yay! we can remove the "don't have wide adoption yet"!!!!! :raised_hands:

File: ch04packaging/03Packaging.ipynb.py [hunk 10]
```diff
@@ -468,16 +507,16 @@ def process():
...
 # ### Define packages and executables
 
 # %% [markdown]
 # We need to create `__init__` files for the source and the tests.
 # ```bash
-# touch greetings/greetings/test/__init__.py
-# touch greetings/greetings/__init__.py
+# touch greetings_repo/tests/__init__.py
+# touch greetings_repo/src/greetings/__init__.py
 # ```
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:520]: we don't need `__init__.py` anymore, and probably the problem with pytest is gone with this file structure as we won't be using relative imports, right?

File: ch04packaging/03Packaging.ipynb.py [hunk 11]
```diff
@@ -489,7 +528,7 @@ def process():
 # Remember, that we need to create the empty `__init__.py` files so that `pytest` can follow the relative imports.
 
 # %% language="bash"
-# touch greetings_repo/greetings/test/__init__.py
+# touch greetings_repo/tests/__init__.py
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:531]: Yeah, probably we can remove all that here too.

File: ch04packaging/03Packaging.ipynb.py [hunk 12]
```diff
@@ -599,19 +646,24 @@ def test_greeter(fixture):
 
 # %% [markdown]
 # Finally, we typically don't want to include the tests when we distribute our software for our users.
-# We can make sure they are not included using the `exclude` option on when telling `setuptools` to find packages.
+# We can make sure they are not included using the `exclude` option on when telling `hatch` to find packages.
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:649]: Do we need the `exclude` option if the tests are not under the package folder?

**Review by dpshelio — CHANGES_REQUESTED** [review]
Amazing job! :rocket:  Love it all, including all the changes to the gh-actions.
I've got a couple of comments about the new file structure and the introduction on package repositories.

File: ch04packaging/03Packaging.ipynb.py (hunk 8 above)
Comment (Saransh-cpp) [reply ch04packaging/03Packaging.ipynb.py:422]: `-e` is introduced at the end of the lesson, but was somehow used before its introduction 😬 

All the other installations blocks after this block and before the introduction of editable installs don't use `-e`.

File: ch04packaging/03Packaging.ipynb.py (hunk 10 above)
Comment (Saransh-cpp) [reply ch04packaging/03Packaging.ipynb.py:524]: Oh yes, we'll need it to add things directly under package's namespace, but definitely not in the tests folder. I'll take it out!

File: ch04packaging/03Packaging.ipynb.py (hunk 12 above)
Comment (Saransh-cpp) [reply ch04packaging/03Packaging.ipynb.py:653]: Ah, not really, I can exclude the `exclude`

File: ch04packaging/03Packaging.ipynb.py [hunk 13]
```diff
@@ -21,43 +21,61 @@
...
+# Besides the guides, there are numerous "cookie" templates available for developers. These templates
+# allow developers to generate an empty Python package following good practices and guidelines with a
+# simple CLI command.
+#  - [Scientific Python cookie](https://github.com/scientific-python/cookie)
+#  - [Open Science Labs' scicookie](https://github.com/osl-incubator/scicookie)
+#  - Domain/ecosystem specific cookies like [pybamm-cookiecutter](https://github.com/pybamm-team/pybamm-cookiecutter) for battery modeling projects in Python exist too
+#
+# Finally, the packaging community regularly organises [PackagingCon](https://packaging-con.org) to discuss the packaging ecosystems
+# of multiple languages and operating systems at a single place.
+#
```
Comment (dpshelio) [inline ch04packaging/03Packaging.ipynb.py:39]: :heart_eyes:

File: ch04packaging/03Packaging.ipynb.py (hunk 7 above)
Comment (Saransh-cpp) [reply ch04packaging/03Packaging.ipynb.py:413]: Missed that, thanks!

File: ch04packaging/03Packaging.ipynb.py [hunk 14]
```diff
@@ -94,30 +113,45 @@
...
+#  - [`flit`](https://flit.pypa.io) allows building pure Python projects with minimal extra configurations
+#  - [`poetry`](https://python-poetry.org) allows building pure Python projects (a full blown dependency and environment management system)
+#  - [`scikit-build-core`](https://scikit-build-core.readthedocs.io) allows building pure Python and Python + C/C++ projects (under active development)
+#  - [`meson`](https://meson-python.readthedocs.io) allows building pure Python and Python + C/C++ projects (has a custom DSL)
+#  - [`maturin`](https://www.maturin.rs): allows building Rust binary extensions
+#
+# In this case, we'll be using `hatch` to build our package, so we list it in the `requires` field. Technically speaking, `hatch` is the front-end (a CLI utility)
+# for the actual build-backend `hatchling`. `hatchling` is installed with hatch and can be specified as the `build-backend` in `pyproject.toml`.
+#
+# Finally, we can set specific options for `hatch` using additional sections in `pyproject.toml`: in this case, we will tell `hatch` that it needs to find **and include** all of the files in our `greetings` folder.
```
Comment (Saransh-cpp) [inline ch04packaging/03Packaging.ipynb.py:132]:
```suggestion
# Finally, we can set specific options for `hatch` using additional sections in `pyproject.toml`: in this case, we will tell `hatch` that it needs to find **and include** all of the files in our `src` folder.
```

File: ch04packaging/03Packaging.ipynb.py (hunk 4 above)
Comment (dpshelio) [reply ch04packaging/03Packaging.ipynb.py:51]: Also probably I'd put higher up: python, julia, r and rust before perl or ruby nowadays.

File: ch04packaging/03Packaging.ipynb.py (hunk 8 above)
Comment (dpshelio) [reply ch04packaging/03Packaging.ipynb.py:423]: Oh, good catch! then I'm happy as it is!!! :raised_hands:

File: ch04packaging/03Packaging.ipynb.py (hunk 12 above)
Comment (dpshelio) [reply ch04packaging/03Packaging.ipynb.py:654]: `exclude` exclude to exclude the exclude.... [buffalo buffalo buffalo...](https://en.wikipedia.org/wiki/Buffalo_buffalo_Buffalo_buffalo_buffalo_buffalo_Buffalo_buffalo)

Comment (Saransh-cpp) [reply ch04packaging/03Packaging.ipynb.py:654]: 😂

**Review by dpshelio — APPROVED** [review]
:raised_hands:

**Pull request merged**